onnxruntime/contrib_ops/cuda/quantization/attention_quantization.cc

```diff
@@ -0,0 +1,135 @@
+      T##_##TQuant,                                                      \
+      kCudaExecutionProvider,                                            \
+      KernelDefBuilder()                                                 \
+          .InputMemoryType<OrtMemTypeCPUInput>(4)                        \
+          .InputMemoryType<OrtMemTypeCPUInput>(5)                        \
+          .InputMemoryType<OrtMemTypeCPUInput>(6)                        \
+          .InputMemoryType<OrtMemTypeCPUInput>(7)                        \
+          .TypeConstraint("T1", DataTypeImpl::GetTensorType<TQuant>())   \
+          .TypeConstraint("T2", DataTypeImpl::GetTensorType<TQuant>())   \
+          .TypeConstraint("T3", DataTypeImpl::GetTensorType<T>())        \
+          .TypeConstraint("T4", DataTypeImpl::GetTensorType<int32_t>()), \
+      QAttention<T, TQuant>);
+
+REGISTER_KERNEL_TYPED(float, int8_t)
+REGISTER_KERNEL_TYPED(MLFloat16, int8_t)
+
+template <typename T>
+Status QAttention<T, int8_t>::ComputeInternal(OpKernelContext* context) const {
+  // Input and output shapes:
+  //   Input 0 - input             : (batch_size, sequence_length, hidden_size)
+  //   Input 1 - weights           : (hidden_size, 3 * hidden_size)
+  //   Input 2 - bias              : (3 * hidden_size)
+  //   Input 3 - mask_index        : (batch_size)
+  //   Input 4 - input_scale       : scalar
+  //   Input 5 - weight_scale      : scalar
+  //   Input 6 - input_zero_point  : scalar
+  //   Input 7 - weight_zero_point : scalar
+  //   Output                      : (batch_size, sequence_length, hidden_size)
+  ORT_RETURN_IF_ERROR(CheckInputs(context));
+  const Tensor* input = context->Input<Tensor>(0);
+  const Tensor* weights = context->Input<Tensor>(1);
+  const Tensor* bias = context->Input<Tensor>(2);
+  const Tensor* mask_index = context->Input<Tensor>(3);
+  const Tensor* input_scale_tensor = context->Input<Tensor>(4);
+  const Tensor* weight_scale_tensor = context->Input<Tensor>(5);
+  /*const Tensor* i_zp_tensor = context->Input<Tensor>(6);
+  const Tensor* w_zp_tensor = context->Input<Tensor>(7);*/
+
+  const auto dims = input->Shape().GetDims();
+  /*int input_size = static_cast<int>(input->Shape().Size());*/
+  int batch_size = static_cast<int>(dims[0]);
+  int sequence_length = static_cast<int>(dims[1]);
+  int hidden_size = static_cast<int>(dims[2]);
+  int head_size = hidden_size / num_heads_;
+
+  TensorShape output_shape(dims);
+  Tensor* output = context->Output(0, output_shape);
+
+  cublasHandle_t cublas = CublasHandle();
+  const size_t element_size = sizeof(T);
+
+  // Use GEMM for fully connection.
+  int m = batch_size * sequence_length;
+  int n = 3 * hidden_size;
+  int k = hidden_size;
+  auto gemm_buffer = GetScratchBuffer<T>(batch_size * sequence_length * 3 * hidden_size * element_size);
+  auto gemm_buffer_quantized = GetScratchBuffer<int32_t>(batch_size * sequence_length * 3 * hidden_size);
```

Comment:
Seems that  gemm_buffer_quantized could be freed after CudaDequantizeWithBias().